### PR TITLE
Fix most Deprecation and Resource warnings.

### PR DESCRIPTION
**Pull request by jamadden**
On Python 3, these made the build output very hard to read (e.g., https://travis-ci.org/buildout/buildout/jobs/394026829)

On Python 3.7, these could actually break the doctests.

Closing the files should get us closer to being able to pass the tests with PyPy.

### CODE REVIEW

**Comment by jamadden** [comment]
The remaining `ResourceWarning`s come from `download.txt` (they don't necessarily show up until later, though). I see them locally and on Travis. A full trace looks like this:
```
/buildout//zope/testrunner/runner.py:406: ResourceWarning: unclosed <socket.socket [closed] fd=10, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6>
  gc.collect()
Object allocated at (most recent call last):
  File "bin/test", lineno 36
  File "//zope/testrunner/__init__.py", lineno 31
    warnings=warnings)
  File "//zope/testrunner/__init__.py", lineno 51
    runner.run()
  File "//zope/testrunner/runner.py", lineno 189
    self.run_tests()
  File "//zope/testrunner/runner.py", lineno 291
    self.skipped, self.import_errors)
  File "//zope/testrunner/runner.py", lineno 460
    import_errors)
  File "//zope/testrunner/runner.py", lineno 383
    test(result)
  File "//python3.7/unittest/case.py", lineno 663
    return self.run(*args, **kwds)
  File "//python3.7/unittest/case.py", lineno 615
    testMethod()
  File "//python3.7/doctest.py", lineno 2193
    test, out=new.write, clear_globs=False)
  File "//python3.7/doctest.py", lineno 1475
    return self.__run(test, compileflags, out)
  File "//python3.7/doctest.py", lineno 1329
    compileflags, 1), test.globs)
  File "<doctest download.txt[131]>", lineno 1
  File "/buildout/src/zc/buildout/download.py", lineno 136
    local_path, is_temp = self.download_cached(url, md5sum)
  File "/buildout/src/zc/buildout/download.py", lineno 167
    _, is_temp = self.download(url, md5sum, cached_path)
  File "/buildout/src/zc/buildout/download.py", lineno 217
    tmp_path, headers = urlretrieve(url, tmp_path)
  File "//python3.7/urllib/request.py", lineno 247
    with contextlib.closing(urlopen(url, data)) as fp:
  File "//python3.7/urllib/request.py", lineno 222
    return opener.open(url, data, timeout)
  File "//python3.7/urllib/request.py", lineno 525
    response = self._open(req, data)
  File "//python3.7/urllib/request.py", lineno 543
    '_open', req)
  File "//python3.7/urllib/request.py", lineno 503
    result = func(*args)
  File "//python3.7/urllib/request.py", lineno 1345
    return self.do_open(http.client.HTTPConnection, req)
  File "//python3.7/urllib/request.py", lineno 1317
    encode_chunked=req.has_header('Transfer-encoding'))
  File "//python3.7/http/client.py", lineno 1229
    self._send_request(method, url, body, headers, encode_chunked)
  File "//python3.7/http/client.py", lineno 1275
    self.endheaders(body, encode_chunked=encode_chunked)
  File "//python3.7/http/client.py", lineno 1224
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "//python3.7/http/client.py", lineno 1016
    self.send(msg)
  File "//python3.7/http/client.py", lineno 956
    self.connect()
  File "//python3.7/http/client.py", lineno 928
    (self.host,self.port), self.timeout, self.source_address)
  File "//python3.7/socket.py", lineno 711
    sock = socket(af, socktype, proto)
```

I *think* what's going on has to do with dual ipv4/ipv6 stacks and the addresses returned from `getaddrinfo` in `socket.create_connection`: it iterates across them until it can connect to one. It's supposedly closing socket objects that fail, but that doesn't seem to be the case, somehow.

I thought that swapping out `"localhost"` to `"127.0.0.1"` in the URLs would fix it, if the hypothesis was correct, but that didn't make a difference.

Still poking...

**Comment by jamadden** [comment]
Nope, it's a bug in `urllib.request.HTTPDefaultErrorHandler`. Unlike `URLopener.http_error_default`, the implementation in that class doesn't close the socket before it raises the exception. Since nothing has been returned yet, the top-level close action taken by `urlretrieve` can't do anything about it.

I think this would be fairly easy to monkey-patch if we cared to (we're already monkey-patching something in urllib.request, but I'm pretty sure it doesn't do anything in Python 3 --- Python 2 has a `urllib._urlopener` attribute, Python 3 has a `urllib.request._opener` attribute of a totally different type)

File: src/zc/buildout/download.py [hunk 1]
```diff
@@ -23,10 +23,20 @@
...
+        def __init__(self, *args, **kwargs):
+            with warnings.catch_warnings():
+                warnings.simplefilter('ignore')
+                # This creates a DeprecationWarning on 3.6:
+                # PatchedURLopener style of invoking requests is deprecated.
+                # Use newer urlopen functions/methods.
+                # Said warning breaks our doctests.
+                super(PatchedURLopener, self).__init__(*args, **kwargs)
+
     request._urlopener = PatchedURLopener()  # Ook! Monkey patch!
```
Comment (jamadden) [inline src/zc/buildout/download.py]: Researching the remaining `ResourceWarnings` has pretty much convinced me that this does nothing in Python 3. I can't find an attribute with that name anywhere in the `urllib` code. There *is* an attribute called `_opener`, which is a different animal; if we try to replace that with this patch, things break badly. 

I'm running tests with this patch completely removed.

**Comment by jamadden** [comment]
The test output seems much cleaner to me. [756 lines now](https://travis-ci.org/buildout/buildout/jobs/398796878) versus [1727 lines before](https://travis-ci.org/buildout/buildout/jobs/399089228). 

Thoughts, @reinout ?

**Comment by reinout** [comment]
Looks good. And the tests still pass, so I'm merging it.

**Comment by reinout** [comment]
Thanks for the cleanup.

**Comment by jamadden** [comment]
Thanks!

**Pull request merged**